### PR TITLE
Obj 321 extract gaz1 date from api response

### PR DESCRIPTION
**Pull request by ICoxCH**
Checks if GAZ1 date should be displayed and if yes makes a call to get it, extract it from the response and display. If not, static text is displayed on the 'confirm' screen.

### CODE REVIEW

**Comment by mwestacott** [comment]
Ready for review but needs to be tested once sub-task https://companieshouse.atlassian.net/browse/OBJ-337 is completed.

File: src/controllers/confirm.company.controller.ts [hunk 1]
```diff
@@ -27,12 +32,24 @@ const INELIGIBLE_PAGES = {
  * @param next
  */
 
-export const get = (req: Request, res: Response, next: NextFunction) => {
+export const get = async (req: Request, res: Response, next: NextFunction) => {
   if (req.session) {
     try {
       const company: ObjectionCompanyProfile = retrieveCompanyProfileFromObjectionSession(req.session);
+
+      const session: Session = req.session as Session;
```
Comment (SVarallo) [inline src/controllers/confirm.company.controller.ts:40]: could this be moved to the top of the get and used in the if?

Comment (mwestacott) [reply src/controllers/confirm.company.controller.ts:40]: I think I see what you mean - move it up and use in the call to retrieve the company profile? Will push a change.

**Review by SVarallo — APPROVED** [review]

File: src/controllers/confirm.company.controller.ts [hunk 2]
```diff
@@ -27,12 +32,25 @@ const INELIGIBLE_PAGES = {
...
-      const company: ObjectionCompanyProfile = retrieveCompanyProfileFromObjectionSession(req.session);
+      const company: ObjectionCompanyProfile = retrieveCompanyProfileFromObjectionSession(session);
+
+      const token: string = retrieveAccessTokenFromSession(session);
+
+      let latestGaz1Date: string;
+      if (await getCompanyEligibility(company.companyNumber, token)) {
+        latestGaz1Date = await getLatestGaz1Date(company.companyNumber, token);
+      } else {
+        latestGaz1Date = "No notice in The Gazette";
```
Comment (markpit) [inline src/controllers/confirm.company.controller.ts:48]: i couldn't see a test to cover the 'else' line 48 scenario

Comment (markpit) [reply src/controllers/confirm.company.controller.ts:48]: i can see it now, it's been added to an existing one

Comment (mwestacott) [reply src/controllers/confirm.company.controller.ts:48]: Yes.

File: src/controllers/confirm.company.controller.ts [hunk 3]
```diff
@@ -69,3 +87,20 @@ export const post = async (req: Request, res: Response, next: NextFunction) => {
 const getIneligiblePage = (apiError: ApiError): string => {
   return INELIGIBLE_PAGES[apiError.data.status];
 };
+
+const getLatestGaz1Date = async (companyNumber: string, token: string): Promise<string> => {
```
Comment (markpit) [inline src/controllers/confirm.company.controller.ts:91]: would this function be better placed in the company.filing.history.service? you could have it return the latest GAZ1 CompanyFilingHistory and then the controller could pick out the date from it. It should make it more re-usable and keep the controller a bit more 'dumb'.

Comment (mwestacott) [reply src/controllers/confirm.company.controller.ts:91]: Yep. Makes sense to me. I'm happy to action this change. @ICoxCH do you agree?

Comment (mwestacott) [reply src/controllers/confirm.company.controller.ts:89]: Changed. Function remains to extract and format the date, but other logic moved to the service.

File: src/services/company.filing.history.service.ts [hunk 4]
```diff
@@ -22,4 +35,8 @@ export const getCompanyFilingHistory = async (companyNumber: string, category: s
   logger.debug("Data from company filing history SDK call " + inspect(sdkResponse));
 
   return sdkResponse.resource as CompanyFilingHistory;
+};
+
+function isGaz1(element: FilingHistoryItem) {
```
Comment (markpit) [inline src/services/company.filing.history.service.ts:40]: can you make this an arrow function to follow suite const ... => etc...

**Review by markpit — APPROVED** [review]

**Pull request merged**